### PR TITLE
add components for new display

### PR DESCRIPTION
**Pull request by jonathan-potter**

### CODE REVIEW

File: src/components/display-wall/grid/DisplayWallGrid.jsx [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+/* removing fontello until the build process across repos is cleaner */
+// import '../../../../fonts/fontello/css/fontello.css'
+import './colors.scss'
+import './DisplayWallGrid.scss'
+
+const ASSETS_PER_GRID = 10
+const NUMBER_OF_ROW_TYPES = 4
+const ROW_HEIGHTS_BY_TYPE = [380, 320, 380, 320]
+const ROW_HEIGHTS_BY_TYPE_MOBILE = [2 * 380, 3 * 320, 2 * 380, 3 * 320]
+const MOBILE_WIDTH_BREAKPOINT = 700
```
Comment (jonathan-potter) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: @harby what should this number be?

File: fonts/fontello/css/animation.css [hunk 2]
```diff
@@ -0,0 +1,85 @@
+/*
+   Animation example, for spinners
+*/
+.animate-spin {
+  -moz-animation: spin 2s infinite linear;
+  -o-animation: spin 2s infinite linear;
+  -webkit-animation: spin 2s infinite linear;
```
Comment (mareksuscak) [inline fonts/fontello/css/animation.css]: We can setup `postcss-loader` with `autoprefixer` to add these vendor prefixes automatically. What do you think?

File: index.js [hunk 3]
```diff
@@ -4,6 +4,8 @@ module.exports = {
   Sidebar: require('./src/components/live-displays/Sidebar/Sidebar'),
   Asset: require('./src/components/live-displays/Asset/Asset'),
   Header: require('./src/components/live-displays/Header/Header'),
+  DisplayWallGrid: require('./src/components/display-wall/grid/DisplayWallGrid'),
+  DisplayWallPaginator: require('./src/paginators/DisplayWallPaginator'),
```
Comment (mareksuscak) [inline index.js]: Possibly set up aliases? You remember when we talked about them in Brazil? Only add aliases for first level folders inside of `src`. No need to do that manually and add something each time you add a new first level folder. We can use webpack's [resolve.fallback](https://webpack.github.io/docs/configuration.html#resolve-fallback).

File: src/Studio.jsx [hunk 4]
```diff
@@ -49,6 +49,6 @@ export default class extends Component {
 
     if (!component) { return null }
 
-    return createElement(component, defaultProps[component.name])
+    return createElement(component, defaultProps[component.displayName || component.name])
```
Comment (mareksuscak) [inline src/Studio.jsx]: I noticed we are repeating this `component.displayName || component.name` a lot here and there. Is there any chance to form a function that does that once? It might not be needed though. Just speculating.

File: src/components/display-wall/grid/DisplayWallGrid.jsx (hunk 1 above)
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: @harby's quiet so I'll add something: http://getbootstrap.com/css/#grid-media-queries

File: src/components/display-wall/grid/DisplayWallGrid.jsx [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+  componentDidMount() {
+    window.addEventListener('resize', this.setWindowWidth)
+  },
+
+  componentDidUnmount() {
+    window.removeEventListener('resize', this.setWindowWidth)
+  },
+
+  setWindowWidth() {
```
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: Please `debounce` this function. Resize is especially that kind of event that when not debounced causes a lot of extra work.

File: src/components/display-wall/grid/DisplayWallGrid.jsx [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    this.setState({
+      windowWidth: window.innerWidth
+    })
+  },
+
+  getInitialState() {
+    return {
+      lightboxActive: false
```
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: There's no initial state of `windowWidth` so the first time we render we use `undefined` value. I'm fine with that but we could invoke `setWindowWidth` in the `componentDidMount` manually to calculate that value.

File: src/components/display-wall/grid/DisplayWallGrid.jsx [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+  render() {
+    const {assets} = this.props
+    const {currentAssetIndex} = this.state
+
+    const assetCloseToEndOfAssets = assets.length - currentAssetIndex < 3
+    if (assetCloseToEndOfAssets) {
+      this.props.loadMoreAssets()
+    }
+
+    let assetsByRow = this.getAssetsByRow()
```
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: `const`

File: src/components/display-wall/grid/DisplayWallGrid.jsx [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+      }
+    })
+
+    setTimeout(() => {
+      /*
+        dispatching a resize event on the window after the render has
+        occured fixes a bug where only a certain number of the assets
+        would be displayed until the user manually resized the window
+      */
+      window.dispatchEvent(new Event('resize'))
```
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: That smells. If all browsers exhibited that behaviour I would suspect that something may be wrong with the code. Could that possibly be related to the comment above where I noticed you don't calculate the initial value of `windowWidth`. It's indeed very suspicious given that after firing `resize` and thus calculating `windowWidth` everything works magically. Hm? Can you verify that?

File: src/components/display-wall/grid/DisplayWallGrid.jsx [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+  renderLightbox() {
+    if (!this.state.lightboxActive) { return }
+
+    let {assets} = this.props
+    let {currentAssetIndex} = this.state
+
+    let currentAsset  = assets[currentAssetIndex]
+    let nextAsset     = assets[currentAssetIndex + 1]
+    let previousAsset = assets[currentAssetIndex - 1]
```
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: these are all `const`

File: src/components/display-wall/grid/DisplayWallGrid.jsx [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    return (
+      <div className="chute-asset__count">
+        <Icon name="heart" className="chute-asset__count-icon" />
+        <span className="chute-asset__count-likes">{asset.hearts}</span>
+      </div>
+    )
+  },
+
+  getAssetsByRow() {
+    let {assets} = this.props
```
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: `const` here and possibly other variables below need to be `const`s as well

File: src/components/display-wall/grid/DisplayWallGrid.jsx [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+
+    let currentAssetIndex = this.props.assets.indexOf(currentAsset)
+
+    this.setState({
+      currentAssetIndex: currentAssetIndex - 1
+    })
+  },
+
+  endEvents() {
+    setTimeout(() => {
```
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: Is `setTimeout` really necessary here?

File: src/components/display-wall/grid/DisplayWallGrid.jsx [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+
+  deactivateLightbox(event) {
+    if (this.eventHandled) { return }
+    this.eventHandled = true
+
+    this.setState({
+      lightboxActive: false
+    })
+
+    let body = document.getElementsByTagName('body')[0]
```
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: Why `document.getElementsByTagName('body')[0]` over just `document.body`?

File: src/components/display-wall/grid/DisplayWallGrid.jsx [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+  onAssetClick(asset, event) {
+    this.setState({
+      currentAssetIndex: asset.index
+    })
+
+    this.activateLightbox(event)
+  },
+
+  onClickCurrentAsset() {
+    this.eventHandled = true
```
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: I don't get the gist of what `eventHandled` does. From what I see it looks like what you really wanted is [stopPropagation](https://developer.mozilla.org/en-US/docs/Web/API/Event/stopPropagation). Can you please explain?

File: src/components/shared/CenteredContentTable.jsx [hunk 14]
```diff
@@ -0,0 +1,35 @@
+'use strict'
```
Comment (mareksuscak) [inline src/components/shared/CenteredContentTable.jsx]: Please remove `'use strict' here and everywhere else. Babel generates that once for the output file.

File: src/components/shared/CenteredContentTable.jsx [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+import React, {PropTypes} from 'react'
+
+import classnames from 'classnames'
+
+const noop = () => {}
```
Comment (mareksuscak) [inline src/components/shared/CenteredContentTable.jsx]: Please consider using lodash noop here and everywhere else. Possibly even upgrading lodash to 4 but if that's not possible, noop in lodash 3.x should be located at `'lodash/lib/noop'` or something like that as far as i remember.

File: src/components/shared/OpenableDropdown.jsx [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    onCloseDropdown: React.PropTypes.func.isRequired,
+    dropdownOpen:  React.PropTypes.bool.isRequired
+  },
+
+  getInitialState() {
+    return {}
+  },
+
+  dropdownClicked(event) {
+    this.setState({ lastEventTarget: event.target })
```
Comment (mareksuscak) [inline src/components/shared/OpenableDropdown.jsx]: From what I see, `lastEventTarget` is not really a state. It's just a helper variable but doesn't affect rendering so I see no reason in using state for storing that kind of info. Use `this.lastEventTarget` instead.

File: src/components/shared/OpenableDropdown.jsx [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+OpenableDropdown.changeOpenMenu = function (newOpenMenu) {
+  if (this.openMenu !== newOpenMenu) {
+    if (this.openMenu) {
+      this.openMenu.closeMenu()
+    }
+
+    this.openMenu = newOpenMenu
+  }
+}
+OpenableDropdown.removeOpenMenu = function (menuToBeClosed) {
```
Comment (mareksuscak) [inline src/components/shared/OpenableDropdown.jsx]: I don't get it why these have to be defined on the class. Looks like they're exclusively used in the instance methods and we're passing in an instance of `this`. Smells.

File: src/util/bootstrapIcons.js [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+export default {
+  'arrow-left':  arrowLeft,
+  'arrow-right': arrowRight,
+  'facebook':    facebook,
+  'heart':       heart,
+  'pinterest':   pinterest,
+  'share':       share,
+  'twitter':     twitter,
+  'x':           x
+}
```
Comment (mareksuscak) [inline src/util/bootstrapIcons.js]: We could utilize [re-exporting](http://www.2ality.com/2014/09/es6-modules-final.html#re-exporting) here.

File: webpack.config.js [hunk 19]
```diff
@@ -39,15 +39,26 @@ module.exports = {
...
-      'live-displays': path.join(__dirname, 'src', 'components', 'live-displays'),
-      'react': path.join(__dirname, 'node_modules', 'react')
+      'cdc':               path.join(__dirname),
+      'cdc-fonts':         path.join(__dirname, 'fonts'),
+      'cdc-icons':         path.join(__dirname, 'icons'),
+      'cdc-src':           path.join(__dirname, 'src'),
+      'cdc-components':    path.join(__dirname, 'src', 'components'),
+      'cdc-live-displays': path.join(__dirname, 'src', 'components', 'live-displays'),
+      'cdc-display-wall':  path.join(__dirname, 'src', 'components', 'display-wall'),
+      'cdc-shared':        path.join(__dirname, 'src', 'components', 'shared'),
```
Comment (mareksuscak) [inline webpack.config.js]: Please no nested aliases when possible. Also use aforementioned resolve.fallback instead of defining each alias manually.

File: src/components/display-wall/grid/DisplayWallLightbox.jsx [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+    /* we might end up making these links but w/e */
+  },
+
+  shareAssetOnFacebook() {
+    /* we might end up making these links but w/e */
+  },
+
+  shareAssetOnTwitter() {
+    /* we might end up making these links but w/e */
+  },
```
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallLightbox.jsx]: These three functions are unused at moment. Do we plan to use them? If so, please add TODOs so we don't forget. That serves as a warning that something is still unfinished.

File: src/components/display-wall/grid/DisplayWallLightbox.jsx [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+
+    return (
+      <button className="c-button c-button--link lightbox__prev" onClick={this.props.onClickPreviousAsset.bind(null, this.props.currentAsset)}>
+        <Icon name="arrow-left" size="large" />
+      </button>
+    )
+  },
+
+  renderNextButton() {
+    if (!this.props.nextAsset) { return }
```
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallLightbox.jsx]: I prefer returning `null` when it comes to rendering functions in React. Because if you return `undefined` from the component `render` function React yells at you that you might've forgotten something. So you have to use `null`. So I use the same approach everywhere because it's clear what it does then.

File: src/components/display-wall/grid/DisplayWallLightbox.jsx [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+      lightboxMaxWidth: 0
+    }
+  },
+
+  preventLightboxDeactivation(e) {
+    e.stopPropagation()
+  },
+
+  listenForEsc: function(event) {
+    if (event.key === 'Escape' || event.keyCode === 27) {
```
Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallLightbox.jsx]: Hah support for `KeyboardEvents.key` seems to be pretty recent in Chrome based on the MDN and caniuse data. In fact it landed in the latest version. It's weird though because it's been supported in IE9+ and FF for a long time now. 

Also it's still a working draft so I'm not sure if we want to use APIs in this stage. What do you think? We could use `keycode` npm package instead.

Source: 
http://caniuse.com/#feat=keyboardevent-key
https://www.w3.org/TR/uievents/#widl-KeyboardEvent-key
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key

File: index.js (hunk 3 above)
Comment (jonathan-potter) [inline index.js]: I agree. I'd rather limit the changes in this PR. definitely something to do in the future

File: fonts/fontello/css/animation.css (hunk 2 above)
Comment (jonathan-potter) [inline fonts/fontello/css/animation.css]: @harby?

File: src/components/display-wall/grid/DisplayWallGrid.jsx (hunk 6 above)
Comment (jonathan-potter) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: good point

File: src/components/display-wall/grid/DisplayWallGrid.jsx (hunk 11 above)
Comment (jonathan-potter) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: yes

File: src/components/display-wall/grid/DisplayWallGrid.jsx (hunk 12 above)
Comment (jonathan-potter) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: because i wasn't aware of `document.body`. changed

File: src/components/shared/OpenableDropdown.jsx (hunk 16 above)
Comment (jonathan-potter) [inline src/components/shared/OpenableDropdown.jsx]: this has been the design of `OpenableDropdown` for over a year. It works in other places. I don't want to change it just for this project

File: src/util/bootstrapIcons.js (hunk 18 above)
Comment (jonathan-potter) [inline src/util/bootstrapIcons.js]: take another look at re-exporting. using that syntax requires things to all come from the same file

File: src/components/display-wall/grid/DisplayWallLightbox.jsx (hunk 20 above)
Comment (jonathan-potter) [inline src/components/display-wall/grid/DisplayWallLightbox.jsx]: it's already pretty obvious that they're stubbed out to be used later

File: src/components/display-wall/grid/DisplayWallLightbox.jsx (hunk 21 above)
Comment (jonathan-potter) [inline src/components/display-wall/grid/DisplayWallLightbox.jsx]: react does not complain here.

My preference is to use `undefined` when possible because when `null` is used I start trying to figure out what requires `null`. In this case nothing requires `null`.

 You are correct that the `render` function for a component will complain about returning an `undefined`. This is not the `render` function.

File: src/components/display-wall/grid/DisplayWallLightbox.jsx (hunk 22 above)
Comment (jonathan-potter) [inline src/components/display-wall/grid/DisplayWallLightbox.jsx]: @harby

File: webpack.config.js (hunk 19 above)
Comment (jonathan-potter) [inline webpack.config.js]: please give me an example of what that would look like and how that will effect aliases.

Comment (mareksuscak) [inline webpack.config.js]: @jonathan-potter you mean `resolve.fallback` example?

File: src/components/display-wall/grid/DisplayWallGrid.jsx (hunk 10 above)
Comment (jonathan-potter) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: this code precedes the widespread use of `const` in our codebases. Switching is an easy thing, but requires a pretty big diff surface area. I'd like to leave this for a cleanup PR.

File: src/Studio.jsx (hunk 4 above)
Comment (jonathan-potter) [inline src/Studio.jsx]: I find one instance of `component.displayName` in the codebase.

Comment (jonathan-potter) [inline src/Studio.jsx]: oh `option`... hurray for shitty naming.

Comment (mareksuscak) [inline src/Studio.jsx]: search for `option.displayName`, forgot to mention, sorry, above this file

Comment (jonathan-potter) [inline src/Studio.jsx]: I'd rather not...

File: fonts/fontello/css/animation.css (hunk 2 above)
Comment (jonathan-potter) [inline fonts/fontello/css/animation.css]: @harby

File: webpack.config.js (hunk 19 above)
Comment (jonathan-potter) [inline webpack.config.js]: yes, it sounds like you know what this will look like with the use of `resolve.fallback`. I'm not sure. an example would be very helpful for me.

Comment (mareksuscak) [inline webpack.config.js]: @jonathan-potter here's a an example: https://github.com/webpack/docs/wiki/troubleshooting#npm-linked-modules-dont-find-their-dependencies - the first line, of course you can define:

```
resolve: { fallback: path.join(__dirname, 'src') }
```

File: src/components/display-wall/grid/DisplayWallGrid.jsx (hunk 13 above)
Comment (jonathan-potter) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: I avoid `stopPropagation`. there are several reasons to do so and I was going to write them, but this article does a great job: https://css-tricks.com/dangers-stopping-event-propagation/

tl;dr: stopping propagation will often do what you want, but if something else is paying attention to that event, you are screwed.

The solution presented in that article is interesting, but isn't one that I've gotten comfortable yet. preventing default is fine, but it doesn't really convey what i'm trying to do.

Comment (mareksuscak) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: gotcha, i still don't like it though cause we are basically reimplementing `stopPropagation` for one time use but w/e

File: src/components/shared/OpenableDropdown.jsx (hunk 17 above)
Comment (jonathan-potter) [inline src/components/shared/OpenableDropdown.jsx]: class variables vs instance variables

File: webpack.config.js (hunk 19 above)
Comment (jonathan-potter) [inline webpack.config.js]: this doesn't sound like it does the same thing. i'm happy to talk about it

File: fonts/fontello/demo.html [hunk 23]
```diff
@@ -0,0 +1,330 @@
+<!DOCTYPE html>
```
Comment (chiefGui) [inline fonts/fontello/demo.html]: Do we need this file at all?

**Comment by chiefGui** [comment]
I don't think this PR makes sense at all. As far as I can comprehend, `DisplayWallGrid` IS NOT A COMPONENT and we're merging this into `chute-display-components`.

`chute-display-components` is what it is: an exposure of components. I think that is more appropriate to create a template called `component-live-simple-wall` or something rather than abstracting everything in this component thing.

Let's pick as an example `component-live-wall`—[it is consuming a `Grid` component](https://github.com/chute/component-live-wall/blob/master/src/Layout.jsx#L3) and then designs its own layout based on that.

I'd like to hear from you guys your thoughts.

**Comment by mareksuscak** [comment]
I'm not a display expert. Don't have a deep knowledge but I understand what you're saying and it makes sense.

File: fonts/fontello/css/animation.css (hunk 2 above)
Comment (harby) [inline fonts/fontello/css/animation.css]: Yeah, that's a good idea! It's always better having autoprefixer.

File: src/components/display-wall/grid/DisplayWallGrid.jsx (hunk 1 above)
Comment (harby) [inline src/components/display-wall/grid/DisplayWallGrid.jsx]: That is correct! Those are the default breakpoints of iota and in this case, I use the default ones!

File: src/components/display-wall/grid/DisplayWallLightbox.jsx (hunk 22 above)
Comment (harby) [inline src/components/display-wall/grid/DisplayWallLightbox.jsx]: Yeah, I see is not supported in Safari at all too. I think this package is a good idea!

File: fonts/fontello/demo.html (hunk 23 above)
Comment (jonathan-potter) [inline fonts/fontello/demo.html]: yes. always.

File: src/components/display-wall/grid/DisplayWallLightbox.jsx (hunk 22 above)
Comment (jonathan-potter) [inline src/components/display-wall/grid/DisplayWallLightbox.jsx]: @harby this part of the code is all yours
